linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala

```diff
@@ -2906,6 +2891,143 @@ private[emitter] class FunctionEmitter(sjsGen: SJSGen) {
+        withDynamicGlobalVar("s", (className, method.name)) { v =>
+          js.Apply(v, newArgs)
+        }
+      }
+
+      val captures = capturesBuilder.result()
+
+      if (captures.isEmpty)
+        innerCall
+      else
+        genIIFE(captures, js.Return(innerCall))
+    }
+
+    private def transformClosure(tree: Closure)(implicit env: Env): js.Tree = {
+      val Closure(arrow, captureParams, params, restParam, body, captureValues) = tree
+
+      implicit val pos = tree.pos
+
+      val capturesBuilder = List.newBuilder[(js.ParamDef, js.Tree)]
+
+      val envVarsForCaptures = (for {
+        (param, value) <- captureParams.zip(captureValues)
+      } yield {
+        assert(!param.mutable, f"Found mutable capture at ${param.pos}")
+
+        val captureName = param.name.name
+
+        val varKind = prepareCapture(value, Some(captureName), arrow) { () =>
+          capturesBuilder += transformParamDef(param) -> transformExpr(value, param.ptpe)
+        }
+
+        captureName -> varKind
+      }).toMap
+
+      val innerFunction = {
+        val bodyEnv = Env.empty(AnyType)
+          .withParams(params)
+          .withVars(envVarsForCaptures)
+
+        desugarToFunctionInternal(arrow, params, restParam, body, isStat = false, bodyEnv)
+      }
+
+      val captures = capturesBuilder.result()
+
+      if (captures.isEmpty) {
+        innerFunction
+      } else {
+        genIIFE(captures, js.Return(innerFunction))
+      }
+    }
+
+    private def prepareCapture(value: Tree, forceName: Option[LocalName], arrow: Boolean)(
+        explicitCapture: () => Unit)(implicit env: Env): VarKind = {
+      def permitImplicitJSThisCapture =
+        esFeatures.useECMAScript2015Semantics && arrow
+
+      value match {
+        case VarRef(name) =>
+          env.varKind(name) match {
+            case VarKind.Immutable if !env.inLoopForVarCapture && forceName.forall(_ == name.name) =>
+              /* forceName is needed when capturing for Closure trees:
+               *
+               * If the name we want to capture implicitly isn't the same as the
+               * capture param name, we cannot capture implicitly: we'd have to
+               * ensure that inside the closure body there is no other var
+               * named like the one we want to capture.
+               * However, that would need a full rename pass.
+               *
+               * Note that with the optimizer enabled, this is unlikely to
+               * ever happen, because the optimizer tries to give the same
+               * names in this case.
+               *
+               * For other usages (notably ApplyDynamicImport), we generate the
+               * body, so it's easy to ensure no collision.
+               */
+              VarKind.Immutable
+
+            case VarKind.ThisAlias if permitImplicitJSThisCapture =>
+              VarKind.ThisAlias
+
+            case VarKind.VarAlias(ident) =>
+              VarKind.VarAlias(ident)
```

Comment:
No, because the VarKind can be bound to any name. Or rather: The binding to the name is not the problem, but the aliasing: Because `VarAlias` must be a FileLevel var, we have a guarantee that no name in the inner scope can shadow it. (otherwise the whole VarAlias mechanism would not work).

Maybe that would be another advantage of `ExplicitThisAlias` (then again `FileLevelVarAlias` would also do the trick).

WDYT? Any suggestions on how to make this clearer?